sst/SimengMemInterface.cc

```diff
@@ -0,0 +1,279 @@
+    */
+    if (addrEnd > max_addr_memory || unsignedOverflow_(addrStart, size)) {
+        completed_read_requests.push_back({target, RegisterValue(), requestId});
+        return;
+    }
+
+    AggregateReadRequest* aggrReq = new AggregateReadRequest(target, requestId);
+    std::vector<StandardMem::Request*> requests = makeSSTRequests<AggregateReadRequest>(aggrReq, addrStart, addrEnd, size);
+    for (StandardMem::Request* req : requests) {
+        mem->send(req);
+    }
+}
+
+void SimengMemInterface::requestWrite(const MemoryAccessTarget& target, const RegisterValue& data) {
+    uint64_t addrStart = target.address;
+    uint64_t size = unsigned(target.size);
+    uint64_t addrEnd = addrStart + size - 1;
+
+    AggregateWriteRequest* aggrReq = new AggregateWriteRequest(target, data);
+    std::vector<StandardMem::Request*> requests = makeSSTRequests<AggregateWriteRequest>(aggrReq, addrStart, addrEnd, size);
+    for (StandardMem::Request* req : requests) {
+        mem->send(req);
+    }
+}
+
+void SimengMemInterface::tick() {
+    tickCounter++;
+}
+
+void SimengMemInterface::clearCompletedReads() {
+    completed_read_requests.clear();
+}
+
+bool SimengMemInterface::hasPendingRequests() const {
+    return completed_read_requests.size() > 0;
+};
+
+const span<MemoryReadResult> SimengMemInterface::getCompletedReads() const {
+    return {const_cast<MemoryReadResult*>(completed_read_requests.data()),
+        completed_read_requests.size()};
+};
+
+void SimengMemInterface::aggregatedReadResponses(AggregateReadRequest* aggrReq) {
+
+    if (aggrReq->aggregateCount != 0) return;
+    std::vector<uint8_t> mergedData;
+    // Loop through the ordered map and merge the data in order inside the mergedData vector.
+    // Also remove entries from the aggregation_map as we loop through each SST Request id.
+    for (auto itr = aggrReq->response_map.begin(); itr != aggrReq->response_map.end(); itr++)
+    {
+        mergedData.insert(mergedData.end(), itr->second.begin(), itr->second.end());
+        aggregation_map.erase(itr->first);
+    }
+    // Send the completed read request back to SimEng via the completed_read_requests
+    // queue.
+    const char* char_data = reinterpret_cast<const char*>(&mergedData[0]);
+    completed_read_requests.push_back({
+        aggrReq->target,
+        RegisterValue(char_data, uint16_t(unsigned(aggrReq->target.size))),
+        aggrReq->id
+    });
+
+    // Cleanup
+    aggrReq->response_map.clear();
+    delete aggrReq;
+}
+
+void SimengMemInterface::SimengMemHandlers::handle(StandardMem::WriteResp* rsp) {
+    delete rsp;
+}
+
+void SimengMemInterface::SimengMemHandlers::handle(StandardMem::ReadResp* rsp) {
+    uint64_t id = rsp->getID();
+    auto data = rsp->data;
+    delete rsp;
+
+    // Upon recieving a response from SST the aggregation_map is used to retrieve
+    // the AggregatedReadRequest the recieved SST response is a part of.     
+    auto itr = mem_interface.aggregation_map.find(id);
```

Comment:
what is `mem_interface`? I assumed `aggregation_map` is a reference to the private member variable defined in 'sst/include/SimengMemInterface.hh'. Why isn't it just `auto itr = aggregation_map.find(id);`?

Reply:
Looking at the header file I kind of get it now. This shows the importance of commenting/"simple code", hopefully, the requested comments in the header file (akin to those seen in existing SimEng header files) will shed light on this. A new user should be able to understand all of this with some prior c++ experience

Reply:
PTAL

Reply:
I think we should action a f2f talk about this regarding the mem_interface/friend class setup. I understand that it must all work but better in-code clarity is required.

Reply:
yeah sure sounds good.